# # Exploratory Data Analysis, Feature Engineering and Modelling

---

1. Import packages
2. Load data
3. Descriptive statistics of data
4. Modelling

---

## 1. Import packages

In [449]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

## 2. Load data

In [450]:
df = pd.read_csv('C:/Users/KahindiE/Desktop/data_science/clean_data_after_eda.csv')
df.head()

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1      4660             0                0  2009-08-21  2016-08-30   
2       544             0                0  2010-04-16  2016-04-16   
3      1584             0                0  2010-03-30  2016-03-30   
4      4425             0              526  2010-01-13  2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  ...  \
0      2015-11-01   2015-06-23               0.00  ...   
1      2009-08-21   2015-08-31             189.95  ...   
2      2010-04-16   2015-04-17              47.96  ...   
3      2010-03-30   2015-03-31             240.04  ...   
4      2010-01-13   2015-03-09             445.75  ...   

   var_6m_price_off_peak_var  var_6m_price_peak_var  \
0                   0.000131           4.100838e-05   
1                   0.000003           1.217891e-03   
2                   0.000004           9.450150e-08   
3                   0.000003           0.000000e+00   
4                   0.000011           2.896760e-06   

   var_6m_price_mid_peak_var  var_6m_price_off_peak_fix  \
0               9.084737e-04                   2.086294   
1               0.000000e+00                   0.009482   
2               0.000000e+00                   0.000000   
3               0.000000e+00                   0.000000   
4               4.860000e-10                   0.000000   

   var_6m_price_peak_fix  var_6m_price_mid_peak_fix var_6m_price_off_peak  \
0              99.530517                  44.235794              2.086425   
1               0.000000                   0.000000              0.009485   
2               0.000000                   0.000000              0.000004   
3               0.000000                   0.000000              0.000003   
4               0.000000                   0.000000              0.000011   

   var_6m_price_peak  var_6m_price_mid_peak  churn  
0       9.953056e+01           4.423670e+01      1  
1       1.217891e-03           0.000000e+00      0  
2       9.450150e-08           0.000000e+00      0  
3       0.000000e+00           0.000000e+00      0  
4       2.896760e-06           4.860000e-10      0  

[5 rows x 44 columns]

In [452]:
price_df= pd.read_csv('C:/Users/KahindiE/Desktop/data_science/price_data.csv')
price_df["price_date"] = pd.to_datetime(price_df["price_date"], format='%Y-%m-%d')
price_df.head()

id price_date  price_off_peak_var  \
0  038af19179925da21a25619c5a24b745 2015-01-01            0.151367   
1  038af19179925da21a25619c5a24b745 2015-02-01            0.151367   
2  038af19179925da21a25619c5a24b745 2015-03-01            0.151367   
3  038af19179925da21a25619c5a24b745 2015-04-01            0.149626   
4  038af19179925da21a25619c5a24b745 2015-05-01            0.149626   

   price_peak_var  price_mid_peak_var  price_off_peak_fix  price_peak_fix  \
0             0.0                 0.0           44.266931             0.0   
1             0.0                 0.0           44.266931             0.0   
2             0.0                 0.0           44.266931             0.0   
3             0.0                 0.0           44.266931             0.0   
4             0.0                 0.0           44.266931             0.0   

   price_mid_peak_fix  
0                 0.0  
1                 0.0  
2                 0.0  
3                 0.0  
4                 0.0

Let's look at the first 3 rows or the last 3 row of both dataframes to see what the data looks like or you can check any number of row.

In [484]:
df.head(3) # 3 rows

id  cons_12m  cons_gas_12m  cons_last_month  \
0  24011ae4ebbe3035111d65fa7c15bc57         0         54946                0   
1  d29c2c54acc38ff3c0614d0a653813dd      4660             0                0   
2  764c75f661154dac3a6c254cd082ea7d       544             0                0   

   date_activ  date_end  date_modif_prod  date_renewal  forecast_cons_12m  \
0      735034    736130           735903        735772               0.00   
1      733640    736206           733640        735841             189.95   
2      733878    736070           733878        735705              47.96   

   forecast_cons_year  ...  offpeak_diff_dec_jan_energy  \
0                   0  ...                    -0.020057   
1                   0  ...                     0.003767   
2                   0  ...                     0.004670   

   offpeak_diff_dec_jan_power  channel_MISSING  \
0                   -3.700961            False   
1                   -0.177779             True   
2                   -0.177779            False   

   channel_ewpakwlliwisiwduibdlfmalxowmwpci  \
0                                     False   
1                                     False   
2                                     False   

   channel_foosdfpfkusacimwkcsosbicdxkicaua  \
0                                      True   
1                                     False   
2                                      True   

   channel_lmkebamcaaclubfxadlmueccxoimlema  \
0                                     False   
1                                     False   
2                                     False   

   channel_usilxuppasemubllopkaafesmlibmsdf  \
0                                     False   
1                                     False   
2                                     False   

   origin_up_kamkkxfxxuwbdslkwifmmcsiusiuosws  \
0                                       False   
1                                        True   
2                                        True   

   origin_up_ldkssxwpmemidmecebumciepifcamkci  \
0                                       False   
1                                       False   
2                                       False   

   origin_up_lxidpiddsbxsbosboudacockeimpuepw  
0                                        True  
1                                       False  
2                                       False  

[3 rows x 53 columns]

In [485]:
price_df.tail() # last 5 rows

id price_date  price_off_peak_var  \
192997  16f51cdc2baa19af0b940ee1b3dd17d5 2015-08-01            0.119916   
192998  16f51cdc2baa19af0b940ee1b3dd17d5 2015-09-01            0.119916   
192999  16f51cdc2baa19af0b940ee1b3dd17d5 2015-10-01            0.119916   
193000  16f51cdc2baa19af0b940ee1b3dd17d5 2015-11-01            0.119916   
193001  16f51cdc2baa19af0b940ee1b3dd17d5 2015-12-01            0.119916   

        price_peak_var  price_mid_peak_var  price_off_peak_fix  \
192997        0.102232            0.076257           40.728885   
192998        0.102232            0.076257           40.728885   
192999        0.102232            0.076257           40.728885   
193000        0.102232            0.076257           40.728885   
193001        0.102232            0.076257           40.728885   

        price_peak_fix  price_mid_peak_fix  year  month  
192997        24.43733           16.291555  2015      8  
192998        24.43733           16.291555  2015      9  
192999        24.43733           16.291555  2015     10  
193000        24.43733           16.291555  2015     11  
193001        24.43733           16.291555  2015     12

---

## 3. Descriptive statistics of data

### Data types

It is useful to first understand the data that you're dealing with along with the data types of each column. The data types may dictate how you transform and engineer features.

In [486]:
df.columns

Index(['id', 'cons_12m', 'cons_gas_12m', 'cons_last_month', 'date_activ',
       'date_end', 'date_modif_prod', 'date_renewal', 'forecast_cons_12m',
       'forecast_cons_year', 'forecast_discount_energy',
       'forecast_meter_rent_12m', 'forecast_price_energy_off_peak',
       'forecast_price_energy_peak', 'forecast_price_pow_off_peak', 'has_gas',
       'imp_cons', 'margin_gross_pow_ele', 'margin_net_pow_ele', 'nb_prod_act',
       'net_margin', 'num_years_antig', 'pow_max',
       'var_year_price_off_peak_var', 'var_year_price_peak_var',
       'var_year_price_mid_peak_var', 'var_year_price_off_peak_fix',
       'var_year_price_peak_fix', 'var_year_price_mid_peak_fix',
       'var_year_price_off_peak', 'var_year_price_peak',
       'var_year_price_mid_peak', 'var_6m_price_off_peak_var',
       'var_6m_price_peak_var', 'var_6m_price_mid_peak_var',
       'var_6m_price_off_peak_fix', 'var_6m_price_peak_fix',
       'var_6m_price_mid_peak_fix', 'var_6m_price_off_peak',
       'var_6m

In [488]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14584 entries, 0 to 14605
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   id                                          14584 non-null  object 
 1   cons_12m                                    14584 non-null  int64  
 2   cons_gas_12m                                14584 non-null  int64  
 3   cons_last_month                             14584 non-null  int64  
 4   date_activ                                  14584 non-null  int64  
 5   date_end                                    14584 non-null  int64  
 6   date_modif_prod                             14584 non-null  int64  
 7   date_renewal                                14584 non-null  int64  
 8   forecast_cons_12m                           14584 non-null  float64
 9   forecast_cons_year                          14584 non-null  int64  
 10  forecast_discou

### Statistics

Now let's look at some statistics about the datasets. We can use the describe() function to get the statistical summary of the data. The describe() function provides us with the following information:

In [490]:
df.dropna(inplace=True)

In [489]:
df.describe()

cons_12m  cons_gas_12m  cons_last_month     date_activ  \
count  1.458400e+04  1.458400e+04     14584.000000   14584.000000   
mean   1.592368e+05  2.804802e+04     16095.545392  734164.886862   
std    5.734890e+05  1.627986e+05     64382.672493     589.379248   
min    0.000000e+00  0.000000e+00         0.000000  731344.000000   
25%    5.680250e+03  0.000000e+00         0.000000  733787.000000   
50%    1.411700e+04  0.000000e+00       793.000000  734200.000000   
75%    4.073750e+04  0.000000e+00      3383.000000  734612.000000   
max    6.207104e+06  4.154590e+06    771203.000000  735477.000000   

            date_end  date_modif_prod   date_renewal  forecast_cons_12m  \
count   14584.000000     14584.000000   14584.000000       14584.000000   
mean   736172.885217    734869.481349  735800.654553        1867.826267   
std       106.926496       922.562199     117.579129        2387.484358   
min    735991.000000    731344.000000  735045.000000           0.000000   
25%    736082.000000    733996.000000  735707.000000         495.025000   
50%    736177.000000    735010.000000  735806.000000        1112.875000   
75%    736268.000000    735765.000000  735900.000000        2399.737500   
max    736493.000000    735992.000000  735991.000000       82902.830000   

       forecast_cons_year  forecast_discount_energy  ...  \
count        14584.000000              14584.000000  ...   
mean          1400.260217                  0.967841  ...   
std           3249.786591                  5.111900  ...   
min              0.000000                  0.000000  ...   
25%              0.000000                  0.000000  ...   
50%            314.000000                  0.000000  ...   
75%           1745.250000                  0.000000  ...   
max         175375.000000                 30.000000  ...   

       var_6m_price_off_peak_fix  var_6m_price_peak_fix  \
count               14584.000000           14584.000000   
mean                    0.924031               1.455625   
std                    18.231988              13.746304   
min                     0.000000               0.000000   
25%                     0.000000               0.000000   
50%                     0.000000               0.000000   
75%                     0.007962               0.000000   
max                   525.988881             399.471115   

       var_6m_price_mid_peak_fix  var_6m_price_off_peak  var_6m_price_peak  \
count               14584.000000           14584.000000       1.458400e+04   
mean                    0.632337               0.924056       1.455658e+00   
std                     5.888683              18.232154       1.374639e+01   
min                     0.000000               0.000000       0.000000e+00   
25%                     0.000000               0.000004       0.000000e+00   
50%                     0.000000               0.000011       9.450150e-08   
75%                     0.000000               0.007973       2.896760e-06   
max                    80.876029             525.994268       3.994725e+02   

       var_6m_price_mid_peak         churn     year  \
count           1.458400e+04  14584.000000  14584.0   
mean            6.323511e-01      0.097161   2015.0   
std             5.888807e+00      0.296187      0.0   
min             0.000000e+00      0.000000   2015.0   
25%             0.000000e+00      0.000000   2015.0   
50%             0.000000e+00      0.000000   2015.0   
75%             4.860000e-10      0.000000   2015.0   
max             8.087765e+01      1.000000   2015.0   

       offpeak_diff_dec_jan_energy  offpeak_diff_dec_jan_power  
count                 14584.000000                14584.000000  
mean                      0.004565                   -0.277289  
std                       0.012053                    1.348347  
min                      -0.168958                  -40.728885  
25%                       0.003520                   -0.177779  
50%                       0.005344                 

In [453]:
# Extract year and month
price_df['year'] = price_df['price_date'].dt.year
price_df['month'] = price_df['price_date'].dt.month

In [454]:
# Convert 'price_date' column to datetime format
price_df['price_date'] = pd.to_datetime(price_df['price_date'])

In [455]:
# Extract year and month from 'price_date' column
price_df['year'] = price_df['price_date'].dt.year
price_df['month'] = price_df['price_date'].dt.month

In [456]:
# Group by id and month to get average monthly prices
monthly_price_by_id = price_df.groupby(['id', 'year', 'month']).agg({
    'price_off_peak_var': 'mean',
    'price_off_peak_fix': 'mean'
}).reset_index()  # Reset index to convert 'id' and 'month' columns to columns

In [457]:
# Filter for December and January prices
dec_prices = monthly_price_by_id [monthly_price ['month'] == 12].rename(columns={
    'price_off_peak_var': 'dec_off_peak_var',
    'price_off_peak_fix': 'dec_off_peak_fix'
}).drop(columns=['month'])

In [458]:
# Filter for January price
jan_prices =  monthly_price_by_id [monthly_price ['month'] == 1].rename(columns={
    'price_off_peak_var': 'jan_off_peak_var',
    'price_off_peak_fix': 'jan_off_peak_fix'
}).drop(columns=['month'])


In [459]:
# Merge December and January prices on id and year
diff = pd.merge(dec_prices, jan_prices, on= ['id', 'year'], suffixes=('_dec', '_jan'))

In [460]:
# Calculate the difference between December and January prices
diff['offpeak_diff_dec_jan_energy'] = diff['jan_off_peak_var'] - diff['dec_off_peak_var']
diff['offpeak_diff_dec_jan_power'] = diff['jan_off_peak_fix'] - diff['dec_off_peak_fix']

In [461]:
# Select relevant columns
diff = diff[['id', 'year', 'offpeak_diff_dec_jan_energy', 'offpeak_diff_dec_jan_power']]
diff.head()

id  year  offpeak_diff_dec_jan_energy  \
0  0002203ffbb812588b632b9e628cc38d  2015                     0.006192   
1  0004351ebdd665e6ee664792efc4fd13  2015                     0.004104   
2  0010bcc39e42b3c2131ed2ce55246e3c  2015                    -0.050443   
3  0010ee3855fdea87602a5b7aba8e42de  2015                     0.010018   
4  00114d74e963e47177db89bc70108537  2015                     0.003994   

   offpeak_diff_dec_jan_power  
0                   -0.162916  
1                   -0.177779  
2                   -1.500000  
3                   -0.162916  
4                    0.000001

In [462]:
#merge the Dataframe
df = pd.merge(df, diff, on=['id'], how='left')
df.head()

id                     channel_sales  \
0  24011ae4ebbe3035111d65fa7c15bc57  foosdfpfkusacimwkcsosbicdxkicaua   
1  d29c2c54acc38ff3c0614d0a653813dd                           MISSING   
2  764c75f661154dac3a6c254cd082ea7d  foosdfpfkusacimwkcsosbicdxkicaua   
3  bba03439a292a1e166f80264c16191cb  lmkebamcaaclubfxadlmueccxoimlema   
4  149d57cf92fc41cf94415803a877cb4b                           MISSING   

   cons_12m  cons_gas_12m  cons_last_month  date_activ    date_end  \
0         0         54946                0  2013-06-15  2016-06-15   
1      4660             0                0  2009-08-21  2016-08-30   
2       544             0                0  2010-04-16  2016-04-16   
3      1584             0                0  2010-03-30  2016-03-30   
4      4425             0              526  2010-01-13  2016-03-07   

  date_modif_prod date_renewal  forecast_cons_12m  ...  \
0      2015-11-01   2015-06-23               0.00  ...   
1      2009-08-21   2015-08-31             189.95  ...   
2      2010-04-16   2015-04-17              47.96  ...   
3      2010-03-30   2015-03-31             240.04  ...   
4      2010-01-13   2015-03-09             445.75  ...   

   var_6m_price_off_peak_fix  var_6m_price_peak_fix  \
0                   2.086294              99.530517   
1                   0.009482               0.000000   
2                   0.000000               0.000000   
3                   0.000000               0.000000   
4                   0.000000               0.000000   

   var_6m_price_mid_peak_fix  var_6m_price_off_peak  var_6m_price_peak  \
0                  44.235794               2.086425       9.953056e+01   
1                   0.000000               0.009485       1.217891e-03   
2                   0.000000               0.000004       9.450150e-08   
3                   0.000000               0.000003       0.000000e+00   
4                   0.000000               0.000011       2.896760e-06   

   var_6m_price_mid_peak churn    year  offpeak_diff_dec_jan_energy  \
0           4.423670e+01     1  2015.0                    -0.020057   
1           0.000000e+00     0  2015.0                     0.003767   
2           0.000000e+00     0  2015.0                     0.004670   
3           0.000000e+00     0  2015.0                     0.004547   
4           4.860000e-10     0  2015.0                     0.006192   

   offpeak_diff_dec_jan_power  
0                   -3.700961  
1                   -0.177779  
2                   -0.177779  
3                   -0.177779  
4                   -0.162916  

[5 rows x 47 columns]

In [463]:
# Rolling average for off-peak fixed price
monthly_price_by_id['rolling_avg_fix'] = monthly_price_by_id.groupby('id')['price_off_peak_fix'].rolling(window=3).mean().reset_index(0, drop =True)
monthly_price_by_id.head()


id  year  month  price_off_peak_var  \
0  0002203ffbb812588b632b9e628cc38d  2015      1            0.126098   
1  0002203ffbb812588b632b9e628cc38d  2015      2            0.126098   
2  0002203ffbb812588b632b9e628cc38d  2015      3            0.128067   
3  0002203ffbb812588b632b9e628cc38d  2015      4            0.128067   
4  0002203ffbb812588b632b9e628cc38d  2015      5            0.128067   

   price_off_peak_fix  rolling_avg_fix  
0           40.565969              NaN  
1           40.565969              NaN  
2           40.728885        40.620275  
3           40.728885        40.674580  
4           40.728885        40.728885

In [464]:
#Standard deviation for off-peak fixed price
monthly_price_by_id['rolling_std_fix'] = monthly_price_by_id.groupby('id')['price_off_peak_fix'].rolling(window=3).std().reset_index(0, drop =True)
monthly_price_by_id.head()

id  year  month  price_off_peak_var  \
0  0002203ffbb812588b632b9e628cc38d  2015      1            0.126098   
1  0002203ffbb812588b632b9e628cc38d  2015      2            0.126098   
2  0002203ffbb812588b632b9e628cc38d  2015      3            0.128067   
3  0002203ffbb812588b632b9e628cc38d  2015      4            0.128067   
4  0002203ffbb812588b632b9e628cc38d  2015      5            0.128067   

   price_off_peak_fix  rolling_avg_fix  rolling_std_fix  
0           40.565969              NaN              NaN  
1           40.565969              NaN              NaN  
2           40.728885        40.620275         0.094059  
3           40.728885        40.674580         0.094059  
4           40.728885        40.728885         0.000000

In [465]:
# Average prices per season
monthly_price_by_id['seasonal_avg_fix'] = monthly_price_by_id.groupby('id')['price_off_peak_fix'].transform(lambda x: x.rolling(window=3, min_periods=1).mean())
monthly_price_by_id.head()

id  year  month  price_off_peak_var  \
0  0002203ffbb812588b632b9e628cc38d  2015      1            0.126098   
1  0002203ffbb812588b632b9e628cc38d  2015      2            0.126098   
2  0002203ffbb812588b632b9e628cc38d  2015      3            0.128067   
3  0002203ffbb812588b632b9e628cc38d  2015      4            0.128067   
4  0002203ffbb812588b632b9e628cc38d  2015      5            0.128067   

   price_off_peak_fix  rolling_avg_fix  rolling_std_fix  seasonal_avg_fix  
0           40.565969              NaN              NaN         40.565969  
1           40.565969              NaN              NaN         40.565969  
2           40.728885        40.620275         0.094059         40.620275  
3           40.728885        40.674580         0.094059         40.674580  
4           40.728885        40.728885         0.000000         40.728885

In [466]:
#caclulate the difference between the seasonal average and the monthly average
monthly_price_by_id['seasonal_diff_fix'] = monthly_price_by_id['seasonal_avg_fix'] - monthly_price_by_id['price_off_peak_fix']
monthly_price_by_id.head()

id  year  month  price_off_peak_var  \
0  0002203ffbb812588b632b9e628cc38d  2015      1            0.126098   
1  0002203ffbb812588b632b9e628cc38d  2015      2            0.126098   
2  0002203ffbb812588b632b9e628cc38d  2015      3            0.128067   
3  0002203ffbb812588b632b9e628cc38d  2015      4            0.128067   
4  0002203ffbb812588b632b9e628cc38d  2015      5            0.128067   

   price_off_peak_fix  rolling_avg_fix  rolling_std_fix  seasonal_avg_fix  \
0           40.565969              NaN              NaN         40.565969   
1           40.565969              NaN              NaN         40.565969   
2           40.728885        40.620275         0.094059         40.620275   
3           40.728885        40.674580         0.094059         40.674580   
4           40.728885        40.728885         0.000000         40.728885   

   seasonal_diff_fix  
0           0.000000  
1           0.000000  
2          -0.108610  
3          -0.054305  
4           0.000000

In [467]:
# Aggregate average prices per period by company
monthly_price_by_id['period_avg_fix'] = monthly_price_by_id.groupby('id')['price_off_peak_fix'].transform(lambda x: x.rolling(window=3, min_periods=1).mean())
monthly_price_by_id.head()

id  year  month  price_off_peak_var  \
0  0002203ffbb812588b632b9e628cc38d  2015      1            0.126098   
1  0002203ffbb812588b632b9e628cc38d  2015      2            0.126098   
2  0002203ffbb812588b632b9e628cc38d  2015      3            0.128067   
3  0002203ffbb812588b632b9e628cc38d  2015      4            0.128067   
4  0002203ffbb812588b632b9e628cc38d  2015      5            0.128067   

   price_off_peak_fix  rolling_avg_fix  rolling_std_fix  seasonal_avg_fix  \
0           40.565969              NaN              NaN         40.565969   
1           40.565969              NaN              NaN         40.565969   
2           40.728885        40.620275         0.094059         40.620275   
3           40.728885        40.674580         0.094059         40.674580   
4           40.728885        40.728885         0.000000         40.728885   

   seasonal_diff_fix  period_avg_fix  
0           0.000000       40.565969  
1           0.000000       40.565969  
2          -0.108610       40.620275  
3          -0.054305       40.674580  
4           0.000000       40.728885

In [468]:
# Aggregate average prices per period by company
mean_prices_by_month = price_df.groupby(['id']).agg({
    'price_off_peak_var': 'mean',
    'price_peak_var': 'mean',
    'price_mid_peak_var': 'mean',
    'price_off_peak_fix': 'mean',
    'price_peak_fix': 'mean',
    'price_mid_peak_fix': 'mean'
}).reset_index()

In [469]:
# Calculate the mean difference between consecutive periods
mean_prices_by_month['off_peak_peak_var_mean_diff'] = mean_prices_by_month['price_off_peak_var'] - mean_prices_by_month['price_peak_var']
mean_prices_by_month['peak_mid_peak_var_mean_diff'] = mean_prices_by_month['price_peak_var'] - mean_prices_by_month['price_mid_peak_var']
mean_prices_by_month['off_peak_mid_peak_var_mean_diff'] = mean_prices_by_month['price_off_peak_var'] - mean_prices_by_month['price_mid_peak_var']
mean_prices_by_month['off_peak_peak_fix_mean_diff'] = mean_prices_by_month['price_off_peak_fix'] - mean_prices_by_month['price_peak_fix']
mean_prices_by_month['peak_mid_peak_fix_mean_diff'] = mean_prices_by_month['price_peak_fix'] - mean_prices_by_month['price_mid_peak_fix']
mean_prices_by_month['off_peak_mid_peak_fix_mean_diff'] = mean_prices_by_month['price_off_peak_fix'] - mean_prices_by_month['price_mid_peak_fix']

In [470]:
df['has_gas'] = df['has_gas'].replace  ({'t': 1, 'f': 0})
df.groupby(['has_gas']).agg({'churn': 'mean'})  

churn
has_gas          
0        0.100544
1        0.081856

In [471]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14606 entries, 0 to 14605
Data columns (total 47 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              14606 non-null  object 
 1   channel_sales                   14606 non-null  object 
 2   cons_12m                        14606 non-null  int64  
 3   cons_gas_12m                    14606 non-null  int64  
 4   cons_last_month                 14606 non-null  int64  
 5   date_activ                      14606 non-null  object 
 6   date_end                        14606 non-null  object 
 7   date_modif_prod                 14606 non-null  object 
 8   date_renewal                    14606 non-null  object 
 9   forecast_cons_12m               14606 non-null  float64
 10  forecast_cons_year              14606 non-null  int64  
 11  forecast_discount_energy        14606 non-null  float64
 12  forecast_meter_rent_12m         

In [472]:
#covert pd to dateTime
df["date_renewal"] = pd.to_datetime(df["date_renewal"], format='%Y-%m-%d')
df["date_activ"] = pd.to_datetime(df["date_activ"], format='%Y-%m-%d')
df["date_end"] = pd.to_datetime(df["date_end"], format='%Y-%m-%d')


In [473]:
# Convert date columns to datetime
date_columns = ['date_activ', 'date_end', 'date_modif_prod', 'date_renewal']
for col in date_columns:
    df[col] = pd.to_datetime(df[col], errors='coerce')
    df[col] = df[col].map(pd.Timestamp.toordinal)

In [474]:
# One-hot encode categorical variables
df = pd.get_dummies(df, columns=['channel_sales'], prefix='channel')
df = df.drop(columns=['channel_sddiedcslfslkckwlfkdpoeeailfpeds', 'channel_epumfxlbckeskwekxbiuasklxalciiuu', 'channel_fixdbufsefwooaasfcxdxadsiekoceaa'])
df.head()

id  cons_12m  cons_gas_12m  cons_last_month  \
0  24011ae4ebbe3035111d65fa7c15bc57         0         54946                0   
1  d29c2c54acc38ff3c0614d0a653813dd      4660             0                0   
2  764c75f661154dac3a6c254cd082ea7d       544             0                0   
3  bba03439a292a1e166f80264c16191cb      1584             0                0   
4  149d57cf92fc41cf94415803a877cb4b      4425             0              526   

   date_activ  date_end  date_modif_prod  date_renewal  forecast_cons_12m  \
0      735034    736130           735903        735772               0.00   
1      733640    736206           733640        735841             189.95   
2      733878    736070           733878        735705              47.96   
3      733861    736053           733861        735688             240.04   
4      733785    736030           733785        735666             445.75   

   forecast_cons_year  ...  var_6m_price_mid_peak  churn    year  \
0                   0  ...           4.423670e+01      1  2015.0   
1                   0  ...           0.000000e+00      0  2015.0   
2                   0  ...           0.000000e+00      0  2015.0   
3                   0  ...           0.000000e+00      0  2015.0   
4                 526  ...           4.860000e-10      0  2015.0   

   offpeak_diff_dec_jan_energy  offpeak_diff_dec_jan_power  channel_MISSING  \
0                    -0.020057                   -3.700961            False   
1                     0.003767                   -0.177779             True   
2                     0.004670                   -0.177779            False   
3                     0.004547                   -0.177779            False   
4                     0.006192                   -0.162916             True   

   channel_ewpakwlliwisiwduibdlfmalxowmwpci  \
0                                     False   
1                                     False   
2                                     False   
3                                     False   
4                                     False   

   channel_foosdfpfkusacimwkcsosbicdxkicaua  \
0                                      True   
1                                     False   
2                                      True   
3                                     False   
4                                     False   

   channel_lmkebamcaaclubfxadlmueccxoimlema  \
0                                     False   
1                                     False   
2                                     False   
3                                      True   
4                                     False   

   channel_usilxuppasemubllopkaafesmlibmsdf  
0                                     False  
1                                     False  
2                                     False  
3                                     False  
4                                     False  

[5 rows x 51 columns]

In [475]:
df = pd.get_dummies(df, columns=['origin_up'], prefix='origin_up')
df = df.drop(columns=['origin_up_MISSING', 'origin_up_usapbepcfoloekilkwsdiboslwaxobdp', 'origin_up_ewxeelcelemmiwuafmddpobolfuxioce'])
df.head()

id  cons_12m  cons_gas_12m  cons_last_month  \
0  24011ae4ebbe3035111d65fa7c15bc57         0         54946                0   
1  d29c2c54acc38ff3c0614d0a653813dd      4660             0                0   
2  764c75f661154dac3a6c254cd082ea7d       544             0                0   
3  bba03439a292a1e166f80264c16191cb      1584             0                0   
4  149d57cf92fc41cf94415803a877cb4b      4425             0              526   

   date_activ  date_end  date_modif_prod  date_renewal  forecast_cons_12m  \
0      735034    736130           735903        735772               0.00   
1      733640    736206           733640        735841             189.95   
2      733878    736070           733878        735705              47.96   
3      733861    736053           733861        735688             240.04   
4      733785    736030           733785        735666             445.75   

   forecast_cons_year  ...  offpeak_diff_dec_jan_energy  \
0                   0  ...                    -0.020057   
1                   0  ...                     0.003767   
2                   0  ...                     0.004670   
3                   0  ...                     0.004547   
4                 526  ...                     0.006192   

   offpeak_diff_dec_jan_power  channel_MISSING  \
0                   -3.700961            False   
1                   -0.177779             True   
2                   -0.177779            False   
3                   -0.177779            False   
4                   -0.162916             True   

   channel_ewpakwlliwisiwduibdlfmalxowmwpci  \
0                                     False   
1                                     False   
2                                     False   
3                                     False   
4                                     False   

   channel_foosdfpfkusacimwkcsosbicdxkicaua  \
0                                      True   
1                                     False   
2                                      True   
3                                     False   
4                                     False   

   channel_lmkebamcaaclubfxadlmueccxoimlema  \
0                                     False   
1                                     False   
2                                     False   
3                                      True   
4                                     False   

   channel_usilxuppasemubllopkaafesmlibmsdf  \
0                                     False   
1                                     False   
2                                     False   
3                                     False   
4                                     False   

   origin_up_kamkkxfxxuwbdslkwifmmcsiusiuosws  \
0                                       False   
1                                        True   
2                                        True   
3                                        True   
4                                        True   

   origin_up_ldkssxwpmemidmecebumciepifcamkci  \
0                                       False   
1                                       False   
2                                       False   
3                                       False   
4                                       False   

   origin_up_lxidpiddsbxsbosboudacockeimpuepw  
0                                        True  
1                                       False  
2                                       False  
3                                       False  
4                                       False  

[5 rows x 53 columns]

### Data sampling

The first thing we want to do is split our dataset into training and test samples. The reason why we do this, is so that we can simulate a real life situation by generating predictions for our test sample, without showing the predictive model these data points. This gives us the ability to see how well our model is able to generalise to new data, which is critical.

A typical % to dedicate to testing is between 20-30, for this example we will use a 75-25% split between train and test respectively.

In [477]:
# Drop the 'id' column as it is not useful for prediction
X = df.drop(['id', 'churn'], axis=1)
y = df['churn']
print(X.shape)
print(y.shape)

(14584, 51)
(14584,)


In [478]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score

In [479]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

(10938, 51)
(10938,)
(3646, 51)
(3646,)


### Model training

Once again, we will using a `Random Forest` classifier in this dataset. A Random Forest sits within the category of `ensemble` algorithms because internally the `Forest` refers to a collection of `Decision Trees` which are tree-based learning algorithms. As the data scientist, you can control how large the forest is (that is, how many decision trees you want to include).

The reason why an `ensemble` algorithm is powerful is because of the laws of averaging, weak learners and the central limit theorem. If we take a single decision tree and give it a sample of data and some parameters, it will learn patterns from the data. It may be overfit or it may be underfit, but that is now our only hope, that single algorithm. 

With `ensemble` methods, instead of banking on 1 single trained model, we can train 1000's of decision trees, all using different splits of the data and learning different patterns. It would be like asking 1000 people to all learn how to code. You would end up with 1000 people with different answers, methods and styles! The weak learner notion applies here too, it has been found that if you train your learners not to overfit, but to learn weak patterns within the data and you have a lot of these weak learners, together they come together to form a highly predictive pool of knowledge! This is a real life application of many brains are better than 1.

Now instead of relying on 1 single decision tree for prediction, the random forest puts it to the overall views of the entire collection of decision trees. Some ensemble algorithms using a voting approach to decide which prediction is best, others using averaging. 

As we increase the number of learners, the idea is that the random forest's performance should converge to its best possible solution.

Some additional advantages of the random forest classifier include:

- The random forest uses a rule-based approach instead of a distance calculation and so features do not need to be scaled
- It is able to handle non-linear parameters better than linear based models

On the flip side, some disadvantages of the random forest classifier include:

- The computational power needed to train a random forest on a large dataset is high, since we need to build a whole ensemble of estimators.
- Training time can be longer due to the increased complexity and size of thee ensemble

In [480]:
# Add model training in here!
# Create an instance of the model
model = RandomForestClassifier()

# Fit the model to the data
model.fit(X, y)

RandomForestClassifier()

In [481]:
# Make predictions on the test set
y_pred = model.predict(X_test)
y_pred_proba = model.predict_proba(X_test)[:, 1]
print(y_pred_proba)

[0.06 0.   0.04 ... 0.02 0.77 0.01]


In [483]:
# Train the classifier
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate the classifier
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

# Print the results
print('Accuracy:', accuracy)
print('Classification Report:', report)

Accuracy: 0.910312671420735
Classification Report:               precision    recall  f1-score   support

           0       0.91      1.00      0.95      3300
           1       0.91      0.06      0.11       346

    accuracy                           0.91      3646
   macro avg       0.91      0.53      0.53      3646
weighted avg       0.91      0.91      0.87      3646



Accuracy: The overall accuracy of your model is 91.03%, meaning it correctly predicted 91.03% of the total cases in your test set.In summary, our model performs well in terms of overall accuracy but struggles with recall for class 1, indicating it may not be effectively identifying instances of this class. Depending on our specific goals and the nature of our  problem, we may need to further tune our model or consider techniques to address class imbalance if relevant.